config/versions/v4_test.go

```diff
@@ -0,0 +1,77 @@
+package versions
+
+import (
+	"bytes"
+	"context"
+	"testing"
+
+	"github.com/buger/jsonparser"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestVersion4ExchangeType(t *testing.T) {
+	t.Parallel()
+	assert.Implements(t, (*ExchangeVersion)(nil), new(Version4))
+}
+
+func TestVersion4Exchanges(t *testing.T) {
+	t.Parallel()
+	assert.Equal(t, []string{"*"}, new(Version4).Exchanges())
+}
+
+func TestVersion4Upgrade(t *testing.T) {
+	t.Parallel()
+
+	_, err := new(Version4).UpgradeExchange(context.Background(), []byte{})
+	require.ErrorIs(t, err, errUpgradingAssetTypes)
+
+	_, err = new(Version4).UpgradeExchange(context.Background(), []byte(`{}`))
+	require.ErrorIs(t, err, errUpgradingCurrencyPairs)
+
+	in := []byte(`{"name":"Cracken","currencyPairs":{"assetTypes":["spot"],"pairs":{"spot":{"enabled":"BTC-AUD","available":"BTC-AUD"},"futures":{"assetEnabled":true},"options":{}}}}`)
```

Comment:
Given that the value in the config can be `"assetEnabled":null` and not just not-present, I think it should part of the test
```
error applying version upgrade to 4: error modifying exchange config: error upgrading currencyPairs.pairs: Value is not a boolean: null
```

Reply:
> I think it should part of the test

Defaults to false, since that's what null would mean.

Fixed https://github.com/gbjk/gocryptotrader/commit/26cb329223

> Speaking of unlabelled errors, which exchange? which asset?

Fixed https://github.com/gbjk/gocryptotrader/commit/0a864c2963
Fixed https://github.com/gbjk/gocryptotrader/commit/65b768612f

Specifically dealt with other data types as well, since that would have silently passed.